hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/scheduler/ContainerScheduler.java

```diff
@@ -0,0 +1,393 @@
+    return this.queuedOpportunisticContainers.size();
+  }
+
+  public OpportunisticContainersStatus getOpportunisticContainersStatus() {
+    this.opportunisticContainersStatus.setQueuedOpportContainers(
+        getNumQueuedOpportunisticContainers());
+    this.opportunisticContainersStatus.setWaitQueueLength(
+        getNumQueuedContainers());
+    return this.opportunisticContainersStatus;
+  }
+
+  private void onContainerCompleted(Container container) {
+    // decrement only if it was a running container
+    if (scheduledToRunContainers.containsKey(container.getContainerId())) {
+      this.utilizationManager.subtractContainerResource(container);
+      if (container.getContainerTokenIdentifier().getExecutionType() ==
+          ExecutionType.OPPORTUNISTIC) {
+        this.opportunisticContainersStatus.setOpportMemoryUsed(
+            this.opportunisticContainersStatus.getOpportMemoryUsed()
+                - container.getResource().getMemorySize());
+        this.opportunisticContainersStatus.setOpportCoresUsed(
+            this.opportunisticContainersStatus.getOpportCoresUsed()
+                - container.getResource().getVirtualCores());
+        this.opportunisticContainersStatus.setRunningOpportContainers(
+            this.opportunisticContainersStatus.getRunningOpportContainers()
+                - 1);
+      }
+    }
+    scheduledToRunContainers.remove(container.getContainerId());
+    oppContainersMarkedForKill.remove(container.getContainerId());
+    startPendingContainers();
+  }
+
+  private void startPendingContainers() {
+    // Start pending guaranteed containers, if resources available.
+    boolean resourcesAvailable =
+        startContainersFromQueue(queuedGuaranteedContainers.values());
+    // Start opportunistic containers, if resources available.
+    if (resourcesAvailable) {
+      startContainersFromQueue(queuedOpportunisticContainers.values());
+    }
+  }
+
+  private boolean startContainersFromQueue(
+      Collection<Container> queuedContainers) {
+    Iterator<Container> cIter = queuedContainers.iterator();
+    boolean resourcesAvailable = true;
+    while (cIter.hasNext() && resourcesAvailable) {
+      Container container = cIter.next();
+      if (this.utilizationManager.hasResourcesAvailable(container)) {
+        startAllocatedContainer(container);
+        cIter.remove();
+      } else {
+        resourcesAvailable = false;
+      }
+    }
+    return resourcesAvailable;
+  }
+
+  @VisibleForTesting
+  protected void scheduleContainer(Container container) {
+    if (maxOppQueueLength <= 0) {
+      startAllocatedContainer(container);
+      return;
+    }
+    if (queuedGuaranteedContainers.isEmpty() &&
+        queuedOpportunisticContainers.isEmpty() &&
+        this.utilizationManager.hasResourcesAvailable(container)) {
+      startAllocatedContainer(container);
+    } else {
+      LOG.info("No available resources for container {} to start its execution "
```

Comment:
So, I felt this should be INFO, since it should happen less frequently, and based on experience from production logs, it is better know immediately why a container did not run (and not have to re-run in DEBUG mode)